File: projects/22-lsm-redis/src/metrics.rs
//! Prometheus metrics for the observability checklist.
//!
//! The [`metrics`] facade writes to a process-global recorder, so the call sites (in
//! [`server`](crate::server), [`engine`](crate::engine), [`wal`](crate::wal),
//! [`compaction`](crate::compaction)) name a metric without knowing about this wiring.
//! Until [`install`] sets a recorder the macros are no-ops — exactly what unit tests
//! want (no setup, no panics).
//!
//! [`install`] returns a [`PrometheusHandle`]; render it from `/metrics` (see
//! [`crate::routes::metrics_router`]). The series the SPEC + boss fight grade:
//! - [`COMMANDS_TOTAL`] — commands served, labelled `cmd` (the ops/sec source).
//! - [`COMMAND_DURATION`] — end-to-end command latency histogram (source of p99).
//! - [`MEMTABLE_BYTES`] — active memtable size (the flush-trigger signal).
//! - [`SSTABLES`] — live SSTable count, labelled `level` (the write-stall backlog signal).
//! - [`COMPACTIONS_TOTAL`] / [`COMPACTION_BYTES_TOTAL`] — compaction progress (V6).
//! - [`BLOCK_CACHE_LOOKUPS_TOTAL`] — labelled `outcome = hit|miss`; `hit / sum` is the ratio.
//! - [`WAL_FSYNC_DURATION`] — fsync latency histogram (the durability/throughput dial, V2).
//! - [`CONNECTED_CLIENTS`] — currently-open RESP connections.
//!
//! Wiring the *call sites* is the observability horizontal item — this module just
//! makes `/metrics` render and single-sources the metric names.

use metrics_exporter_prometheus::{PrometheusBuilder, PrometheusHandle};

/// Commands served, labelled `cmd = get|set|del|…`. Rate of this is ops/sec.
pub const COMMANDS_TOTAL: &str = "lsm_commands_total";

/// End-to-end command latency, in seconds — rendered as quantiles, so p99 falls out.
pub const COMMAND_DURATION: &str = "lsm_command_duration_seconds";

/// Active memtable size in bytes — climbs on writes, drops to ~0 on a flush.
pub const MEMTABLE_BYTES: &str = "lsm_memtable_bytes";

/// Live SSTables, labelled by `level`. A youngest-level count that keeps climbing is
/// the compaction-can't-keep-up (write-stall) signal the boss fight watches.
pub const SSTABLES: &str = "lsm_sstables";

/// Compactions completed (V6).
pub const COMPACTIONS_TOTAL: &str = "lsm_compactions_total";

/// Bytes rewritten by compaction — the write-amplification meter.
pub const COMPACTION_BYTES_TOTAL: &str = "lsm_compaction_bytes_total";

/// Block-cache lookups, labelled `outcome = hit|miss`. Hit ratio = `hit / (hit+miss)`.
pub const BLOCK_CACHE_LOOKUPS_TOTAL: &str = "lsm_block_cache_lookups_total";

/// WAL fsync latency, seconds — the cost the durability policy (V2) is trading against.
pub const WAL_FSYNC_DURATION: &str = "lsm_wal_fsync_duration_seconds";

/// Currently-open RESP connections.
pub const CONNECTED_CLIENTS: &str = "lsm_connected_clients";

/// Install the process-global Prometheus recorder and return a handle used to render
/// the registry for `/metrics`. Call once, from `main`, after telemetry init. Panics
/// if a recorder is already installed (calling it twice is a bug).
pub fn install() -> PrometheusHandle {
    register_descriptions();
    PrometheusBuilder::new()
        .install_recorder()
        .expect("install prometheus recorder")
}

/// Register HELP metadata for the graded metrics. Naming each constant here also gives
/// rustc a direct use of it (proc-macro call sites alone don't always satisfy
/// `dead_code`), so the scaffold stays warning-quiet before the call sites exist.
fn register_descriptions() {
    metrics::describe_counter!(COMMANDS_TOTAL, "Commands served, labelled by cmd");
    metrics::describe_histogram!(COMMAND_DURATION, "End-to-end command latency, seconds");
    metrics::describe_gauge!(MEMTABLE_BYTES, "Active memtable size in bytes");
    metrics::describe_gauge!(SSTABLES, "Live SSTables, labelled by level");
    metrics::describe_counter!(COMPACTIONS_TOTAL, "Compactions completed");
    metrics::describe_counter!(COMPACTION_BYTES_TOTAL, "Bytes rewritten by compaction");
    metrics::describe_counter!(
        BLOCK_CACHE_LOOKUPS_TOTAL,
        "Block-cache lookups, labelled hit|miss"
    );
    metrics::describe_histogram!(WAL_FSYNC_DURATION, "WAL fsync latency, seconds");
    metrics::describe_gauge!(CONNECTED_CLIENTS, "Currently-open RESP connections");
}

File: projects/22-lsm-redis/src/lsm_redis/metrics.py
"""Prometheus metrics for the observability checklist.

`prometheus_client` collectors register themselves into the default registry at
import time, and `common_telemetry.metrics_routes()` renders that same registry
at `/metrics` — so declaring a metric here is all the wiring there is. That
replaces the Rust side's install-a-global-recorder step entirely: no `install()`
to call, no ordering constraint against telemetry init, and no window at startup
where a metric is silently dropped.

The series the SPEC grades, and what each one is *for* during the boss fight:

* `COMMANDS_TOTAL` — labelled `cmd`; its rate is ops/sec by command, which is
  how you tell "SETs collapsed" from "everything collapsed".
* `COMMAND_DURATION` — a histogram, so p99 falls out at query time. See the
  bucket note below; the default buckets cannot answer this SPEC's question.
* `MEMTABLE_BYTES` — climbs on writes, drops to ~0 on flush. A sawtooth means
  flushing works; a ramp means it does not.
* `SSTABLES` — labelled `level`. **The write-stall signal.** A youngest-level
  count that keeps climbing while writes flow is compaction losing, and it
  climbs long before throughput visibly falls, which is what makes it worth
  watching rather than inferring.
* `COMPACTIONS_TOTAL` / `COMPACTION_BYTES_TOTAL` — compaction progress and the
  write-amplification meter (bytes rewritten ÷ bytes written by clients).
* `BLOCK_CACHE_LOOKUPS_TOTAL` — labelled `outcome`; hit ratio is
  `hit / (hit + miss)` at query time.
* `WAL_FSYNC_DURATION` — what the durability policy actually costs, in seconds.
* `CONNECTED_CLIENTS` — open RESP connections.

Two conventions worth copying rather than re-deriving:

**Counters per outcome, never a pre-computed ratio.** There is no
`block_cache_hit_ratio` gauge here on purpose. A ratio cannot be aggregated
across replicas or re-windowed after the fact; two counters can, and Prometheus
divides at query time. The same argument is why `COMMANDS_TOTAL` is labelled by
command rather than split into one metric per command.

**Buckets are a decision.** `prometheus_client`'s default histogram buckets
top out at 10 seconds with nothing between 1 ms and 10 ms — and the SPEC's
target is "p99 ≤ 10 ms", so the defaults would let you claim it while sitting
anywhere in a 9 ms-wide bucket. The buckets below are chosen to be dense exactly
where the target is. A histogram whose buckets do not bracket your SLO is a
metric that cannot fail.

Wiring the *call sites* — bumping these from `server.py`, `engine.py`, `wal.py`
and `compaction.py` — is the observability horizontal item, and is left for you.
This module just declares them and single-sources the names.
"""

from __future__ import annotations

from prometheus_client import Counter, Gauge, Histogram

__all__ = [
    "BLOCK_CACHE_LOOKUPS_TOTAL",
    "COMMANDS_TOTAL",
    "COMMAND_DURATION",
    "COMPACTIONS_TOTAL",
    "COMPACTION_BYTES_TOTAL",
    "CONNECTED_CLIENTS",
    "MEMTABLE_BYTES",
    "SSTABLES",
    "WAL_FSYNC_DURATION",
]

COMMAND_BUCKETS = (
    0.0001,
    0.00025,
    0.0005,
    0.001,
    0.0025,
    0.005,
    0.0075,
    0.01,
    0.025,
    0.05,
    0.1,
    0.25,
    1.0,
)
"""Seconds. Dense from 100 µs (an in-memory hit) to 10 ms (the SPEC's p99
target), then coarse — past 100 ms you only need to know *that* it happened."""

FSYNC_BUCKETS = (0.0001, 0.0005, 0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1.0)
"""Seconds. A page-cache-only "fsync" lands under 100 µs and a real one on an
SSD lands around 1 ms — which is exactly how this histogram tells you whether
`WAL_SYNC=always` is doing what you think it is."""

# ---- Counters (rates) ------------------------------------------------------

COMMANDS_TOTAL = Counter(
    "lsm_commands_total",
    "Commands served, labelled by command name; the rate of this is ops/sec",
    ["cmd"],
)

COMPACTIONS_TOTAL = Counter(
    "lsm_compactions_total",
    "Compactions completed (V6)",
)

COMPACTION_BYTES_TOTAL = Counter(
    "lsm_compaction_bytes_total",
    "Bytes rewritten by compaction; against bytes written by clients this is write amplification",
)

BLOCK_CACHE_LOOKUPS_TOTAL = Counter(
    "lsm_block_cache_lookups_total",
    "Block-cache lookups; hit ratio = hit / (hit + miss)",
    ["outcome"],
)

# ---- Histograms (distributions) --------------------------------------------

COMMAND_DURATION = Histogram(
    "lsm_command_duration_seconds",
    "End-to-end command latency",
    ["cmd"],
    buckets=COMMAND_BUCKETS,
)

WAL_FSYNC_DURATION = Histogram(
    "lsm_wal_fsync_duration_seconds",
    "WAL fsync latency — what the durability policy costs (V2)",
    buckets=FSYNC_BUCKETS,
)

# ---- Gauges (current state) ------------------------------------------------

MEMTABLE_BYTES = Gauge(
    "lsm_memtable_bytes",
    "Approximate bytes held by the active memtable",
)

SSTABLES = Gauge(
    "lsm_sstables",
    "Live SSTables by level; a youngest-level count that keeps climbing is the write stall",
    ["level"],
)

CONNECTED_CLIENTS = Gauge(
    "lsm_connected_clients",
    "Currently-open RESP connections",
)
